### ISSUE TITLE
Qualitative reproduction issues: unstable object reconstruction and hand-object relative pose

### ISSUE THREAD

**Issue by jianian0318** [aidilayce/FollowMyHold#5]
Hi, thanks a lot for the great work and for releasing the code. I was able to set up the environment successfully and start running the pipeline. I am currently doing a qualitative reproduction and found the following issues:

**1. Unreasonable object reconstruction**
The reconstructed object can contain large unrealistic volumes/artifacts.
I checked the preprocessing results and the cropped images and masks seem normal. However, the generated object may still have large implausible geometry. After setting `FOHO_DEBUG_DIR`, I found that this issue already appears in the early object generation stage (t0–t8). I am not sure whether this comes from Hunyuan3D itself or from the current integration. I attach some example images below.
| <img width="100" height="140" alt="Image" src="https://github.com/user-attachments/assets/68fa0e47-0198-4204-8542-7d4679afeb5c" /> | <img width="200" height="100" alt="Image" src="https://github.com/user-attachments/assets/5c5d7bb4-1fc6-402c-a220-91a37b6da7eb" /> |<img width="200" height="100" alt="Image" src="https://github.com/user-attachments/assets/e0ef03dc-308f-42cb-b0df-f2483e7b5a5f" />|
| :---: | :---: | :---: |
|<img width="100" height="1706" alt="Image" src="https://github.com/user-attachments/assets/61b0900e-4e15-44be-85ae-b326ca734931" />|<img width="200" alt="Image" src="https://github.com/user-attachments/assets/204bd068-6991-46ad-924c-401a235c0860" />|<img width="200"  alt="Image" src="https://github.com/user-attachments/assets/54b06e67-45a7-4cd0-99c6-88fa53c1a8a1" />|
| input image  | rendered_normal_t9.png (before optimaztion) | rendered_normal_t19.png (final result) |

**2. Inaccurate hand-object relative pose after optimization**
The optimized relative pose between the hand and the object can be inaccurate. I observed this both on my own examples and also on the teaser case. In addition, the result seems unstable. For the same input image, even with export `FOHO_DETERMINISTIC=1`, two runs can still produce noticeably different results.
For the teaser example in particular, with the same settings, sometimes the result is very poor, while other times it is much closer to the teaser result. I attach comparison examples below.

|<img width="200" alt="Image" src="https://github.com/user-attachments/assets/9667fea4-c859-4f1f-8c81-8767a156ceb4" />|<img width="200" alt="Image" src="https://github.com/user-attachments/assets/a8a738eb-e677-4d5f-882f-ec8557098ff5" />|
| :---: | :---: |
|rendered_normal_t19.png (bad) |rendered_normal_t19.png (good)|

**What I have checked**
1. The environment matches the required setup.
2. Preprocessing outputs exist and look correct, including object masks, cropped images, etc.
3. I tried changing:
`obj_guidance_scale` (default 5.0, I tested 5.0 ~ 9.0)
`guidance_start_step` (default 10, I tested 10 ~ 14)
but the above issues still remain.

I would appreciate any suggestions on whether this is expected behavior, a known issue, or if there are recommended settings for more stable reproduction.

Thanks again!

**Comment by jianian0318** [comment]
I have identified the root cause of the first issue: the Hunyuan `rembg` tool used in the pipeline may sometimes retain background objects instead of removing them. An additional round of object segmentation and background removal should fix the problem.
<img width="1508" height="569" alt="Image" src="https://github.com/user-attachments/assets/1b01b6e2-bffe-4a0c-99dc-17f474d223c8" />

**Comment by aidilayce** [comment]
Hi @jianian0318, thanks for your interest and the detailed issue!

**1.** You’re right that this comes from preprocessing, and thanks for pointing it out. I had not tested this case before, although in hindsight it is quite obvious.

For reference, I’m still sharing the examples below, since they may be useful for documenting the failure mode and a possible workaround. In these cases, the inpainted image still contains the keyboard in addition to the manipulated object as you pointed out due to rembg, which seems to hurt the guidance step and leads to unreasonable meshes. Additional preprocessing step to further mask the inpainted image to isolate only the object using the Gemini-estimated object prompt (with for instance SAM3), then using the masked image during the guidance step works better. After doing these, I can get:

| input image | masked inpainted image| final result |
|------------|---------------------|--------------|
| <img width="150" height="150" alt="Image" src="https://github.com/user-attachments/assets/60e938cf-d5e4-49e0-a08c-ef58bb5bd439" /> | <img width="150" height="150" alt="Image" src="https://github.com/user-attachments/assets/e7afd407-bebb-4686-9c28-7e951195590d" /> | <img width="150" height="150" alt="Image" src="https://github.com/user-attachments/assets/38894ca2-29a0-48f5-a660-793814fa7501" /> |
|<img width="150" height="150" alt="Image" src="https://github.com/user-attachments/assets/18f379df-15f4-4e1a-964e-9867f302d4b8" /> | <img width="150" height="150" alt="Image" src="https://github.com/user-attachments/assets/2f22e8de-7e43-434f-9b56-d41452cba96d" /> |<img width="150" height="150" alt="Image" src="https://github.com/user-attachments/assets/8d76e20d-16b6-4a72-9d95-df06a9d8bf70" /> |

However, even with this, the method does not reconstruct the white box grasped example well. Trying other learning rates or number of optimization steps for Phase 3 (joint optimization) might be helpful.

**2.** For the second part, this is generally related to determinism limitations. As far as I know, some CUDA-based operations in the pipeline are inherently nondeterministic. Specifically, the backward pass of the pytorch3d rasterizer has no deterministic implementation (the related discussion can be found in [the pytorch3d issues](https://github.com/facebookresearch/pytorch3d/issues/1519)). So even with the determinism flag, the main optimization loop can be nondeterministic. Unfortunately, I’m not sure what else can be done except changing the seed or repeating the experiments (which both practices are widely used in the literature as well) to get reasonable-looking reconstructions. 

If you happen to have insights or better solutions for improving determinism here, I’d be very interested to hear them!
